File: tests/performance/protected.lua
math.randomseed(os.time())

local function b64(data)
    local b = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/'
    local res = ((data:gsub('.', function(x)
        local r, b = '', x:byte()
        for i = 8, 1, -1 do r = r .. (b % 2 ^ i - b % 2 ^ (i - 1) > 0 and '1' or '0') end
        return r;
    end) .. '0000'):gsub('%d%d%d?%d?%d?%d?', function(x)
        if (#x < 6) then return '' end
        local c = 0
        for i = 1, 6 do c = c + (x:sub(i, i) == '1' and 2 ^ (6 - i) or 0) end
        return b:sub(c + 1, c + 1)
    end))
    res = res:gsub('%+', '-'):gsub('/', '_'):gsub('=', '')
    return res
end

local tokens = {}
local token_count = 9000

for i = 1, token_count do
    local id = tostring(1000 + i)
    local payload = '{"sub":"' .. id .. '"}'
    tokens[i] = "Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9." .. b64(payload) .. "."
end

wrk.method = "POST"

request = function()
    local token = tokens[math.random(1, token_count)]

    wrk.headers["authorization"] = token
    wrk.headers["x-forwarded-host"] = "that-limit.com"
    return wrk.format()
end

File: tests/performance/protected.js
import http from 'k6/http';
import { check } from 'k6';
import encoding from 'k6/encoding';

const tokenPool = Array.from({ length: 9000 }, (_, i) => {
  const userId = 1000 + i;
  const token = generateFakeToken(userId)

  return `Bearer ${token}`;
});

export const options = {
  stages: [
    { duration: '30s', target: 100 },
    { duration: '1m', target: 100 },
    { duration: '30s', target: 0 },
  ],
  thresholds: {
    http_req_failed: ['rate<0.01'],
    http_req_duration: ['p(99)<10'],
  },
};

export default function () {
  const token = tokenPool[Math.floor(Math.random() * tokenPool.length)];

  const url = 'http://127.0.0.1:8000/consume';

  const params = {
    headers: {
      'x-forwarded-host': 'that-limit.com',
      'authorization': token,
    },
  };

  const res = http.post(url, null, params);

  check(res, {
    'is status 200 or 429': (r) => r.status === 200 || r.status === 429,
  });
}

function generateFakeToken(userId) {
  const header = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9";
  const payload = encoding.b64encode(JSON.stringify({
    sub: userId.toString(),
  }), 'rawurl');

  return `${header}.${payload}.`;
}
